### PR TITLE
Refactor database-fetcher.ts into a class and simplify

### PR DESCRIPTION
**Pull request by robertbrignull**
This PR tries to make the `database-fetcher.ts` file easier to work with, and also sets up for future refactoring in this area.

I highly recommend ignoring whitespace when reviewing! It has also been split into commits if you find that makes it easier.

The main change is creating the `DatabaseFetcher` class, rather than having all functions be defined at the top level and exported. My reasoning for doing this is:
- There's a lot of common arguments to methods. They all take some combination of `app` (or `commandManager`), `databaseManager`, `storagePath`, and `cli`. These are values that don't change and are needed by every method, so it makes sense to pass them in once when constructing the `DatabaseFetcher` object.
- It makes mocking easier because you can use `mockedObject` and pass that in, rather than having to know exactly which methods the code calls and spying on them. It should be a lot less brittle to code changes now, and safer because it'll fail with an error message instead of calling unexpected code.

I then do a little bit of refactoring around the skeleton query wizard because I spotted an opportunity to delete a method and reuse code.

This PR is also hopefully setting up for more work in this area. Right now I haven't changed much about what is done in `database-fetcher.ts`, but my next plan is to maybe separate it into two areas. The idea would be to separate the fetching of databases from a known source (i.e. URL or github NWO) and prompting for which database to import, which is currently spread between `DatabaseFetcher` and `DatabaseUI` and `src/databases/github-databases/download.ts`.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/databases/database-fetcher.ts [hunk 1]
```diff
@@ -39,572 +38,543 @@
...
+      dispose: disposeTimeout,
+    } = createTimeoutSignal(downloadTimeout());
+
+    let response: Response;
+    try {
+      response = await this.checkForFailingResponse(
+        await fetch(databaseUrl, {
+          headers: requestHeaders,
+          signal,
+        }),
```
[the remark this answers is not in the document]
Comment (robertbrignull) [reply extensions/ql-vscode/src/databases/database-fetcher.ts:470]: I've dismissed this alert as a false positive. There's nothing wrong that I can see with the `checkForFailingResponse` function or anything else we do with this response.

I'm not sure why it wasn't present on the main branch already as this is just code that's been moved around.

**Comment by robertbrignull** [comment]
This had some fairly complex merge conflicts and I realised a bunch of the tests were broken. Since nobody has reviewed yet I've tried to rebase and keep all the commits clean. Hopefully I got everything right while rebasing 🤞🏼

File: extensions/ql-vscode/test/vscode-tests/cli-integration/local-queries/skeleton-query-wizard.test.ts [hunk 2]
```diff
@@ -55,12 +56,7 @@ describe("SkeletonQueryWizard", () => {
   let createExampleQlFileSpy: jest.SpiedFunction<
     typeof QlPackGenerator.prototype.createExampleQlFile
   >;
-  let downloadGitHubDatabaseSpy: jest.SpiedFunction<
-    typeof databaseFetcher.downloadGitHubDatabase
-  >;
-  let askForGitHubRepoSpy: jest.SpiedFunction<
-    typeof databaseFetcher.askForGitHubRepo
-  >;
+  let promptImportGithubDatabaseMock: jest.Mock<DatabaseItem | undefined>;
```
Comment (koesie10) [inline extensions/ql-vscode/test/vscode-tests/cli-integration/local-queries/skeleton-query-wizard.test.ts:59]:
```suggestion
  let promptImportGithubDatabaseMock: jest.MockedFunction<
    DatabaseFetcher["promptImportGithubDatabase"]
  >;
```

File: extensions/ql-vscode/test/vscode-tests/no-workspace/databases/github-databases/download.test.ts [hunk 3]
```diff
@@ -116,14 +113,15 @@ describe("downloadDatabaseFromGitHub", () => {
   ];
 
   let showQuickPickSpy: jest.SpiedFunction<typeof window.showQuickPick>;
-  let downloadGitHubDatabaseFromUrlSpy: jest.SpiedFunction<
-    typeof databaseFetcher.downloadGitHubDatabaseFromUrl
-  >;
+  let downloadGitHubDatabaseFromUrlMock: jest.Mock<DatabaseItem | undefined>;
```
Comment (koesie10) [inline extensions/ql-vscode/test/vscode-tests/no-workspace/databases/github-databases/download.test.ts:116]:
```suggestion
  let downloadGitHubDatabaseFromUrlMock: jest.MockedFunction<
    DatabaseFetcher["downloadGitHubDatabaseFromUrl"]
  >;
```

File: extensions/ql-vscode/test/vscode-tests/no-workspace/databases/github-databases/updates.test.ts [hunk 4]
```diff
@@ -367,16 +368,18 @@ describe("downloadDatabaseUpdateFromGitHub", () => {
   ];
 
   let showQuickPickSpy: jest.SpiedFunction<typeof window.showQuickPick>;
-  let downloadGitHubDatabaseFromUrlSpy: jest.SpiedFunction<
-    typeof databaseFetcher.downloadGitHubDatabaseFromUrl
-  >;
+  let downloadGitHubDatabaseFromUrlMock: jest.Mock<DatabaseItem | undefined>;
```
Comment (koesie10) [inline extensions/ql-vscode/test/vscode-tests/no-workspace/databases/github-databases/updates.test.ts:371]:
```suggestion
  let downloadGitHubDatabaseFromUrlMock: jest.MockedFunction<
    DatabaseFetcher["downloadGitHubDatabaseFromUrl"]
  >;
```

File: extensions/ql-vscode/src/databases/database-fetcher.ts [hunk 5]
```diff
@@ -42,605 +41,566 @@ import { createFilenameFromString } from "../common/filenames";
...
+   *
+   * @param progress the progress callback
+   * @param language the language to download. If undefined, the user will be prompted to choose a language.
+   * @param makeSelected make the new database selected in the databases panel (default: true)
+   * @param addSourceArchiveFolder whether to add a workspace folder containing the source archive to the workspace
+   */
+  public async promptImportGithubDatabase(
+    progress: ProgressCallback,
+    language?: string,
+    suggestedRepoNwo?: string,
```
Comment (koesie10) [inline extensions/ql-vscode/src/databases/database-fetcher.ts:111]: The JSDoc above is missing a comment for this parameter, which we should probably add to be consistent.

File: extensions/ql-vscode/src/databases/database-fetcher.ts [hunk 6]
```diff
@@ -42,605 +41,566 @@ import { createFilenameFromString } from "../common/filenames";
...
+  }
+
+  /**
+   * Imports a database from a local archive.
+   *
+   * @param databaseUrl the file url of the archive to import
+   */
+  public async importArchiveDatabase(
+    databaseUrl: string,
+    progress: ProgressCallback,
```
Comment (koesie10) [inline extensions/ql-vscode/src/databases/database-fetcher.ts:276]: This isn't from your PR, but a JSDoc description is missing for `progress`. Could you add this here?

**Review by koesie10 — APPROVED** [review]
LGTM, but this will have quite some conflicts with https://github.com/github/vscode-codeql/pull/3433.

**Comment by robertbrignull** [comment]
Thanks. I'll address your comments but then wait for https://github.com/github/vscode-codeql/pull/3433 to be merged so I can fix up the conflicts manually. That's probably easier than fixing them on that PR.

**Comment by robertbrignull** [comment]
I've merged in latest `main`. There were conflicts in:
- `src/databases/database-fetcher.ts`
- `src/databases/local-databases-ui.ts`
- `test/vscode-tests/cli-integration/databases/database-fetcher.test.ts`
- `test/vscode-tests/global.helper.ts`

I fixed the conflicts largely by manually re-applying the changes from https://github.com/github/vscode-codeql/pull/3433/files and https://github.com/github/vscode-codeql/pull/3486/files.

The only change that wasn't clear how to fix was around the `ensureZippedSourceLocation` function which now gets used from `DatabaseManager`. I got around this by moving it to a different file instead of keeping it as a method of `DatabaseFetcher`.

It's quite hard to see everything that changed during the merge so this may need another full review if you want to be certain. To help with this you can compare to https://github.com/github/vscode-codeql/compare/66de756d8...ff889b73e10af3dd3889acb946060cf722a3df53 which is the diff before the merge.

File: extensions/ql-vscode/src/databases/local-databases/database-contents.ts [hunk 7]
```diff
@@ -1,4 +1,7 @@
+import { pathExists, remove } from "fs-extra";
+import { join } from "path/posix";
```
Comment (koesie10) [inline extensions/ql-vscode/src/databases/local-databases/database-contents.ts:2]: The `join` in `database-fetcher.ts` seems to come from `path`, not `path/posix`:

```suggestion
import { join } from "path";
```

**Review by koesie10 — APPROVED** [review]
Thanks, LGTM. I found two minor unexpected changes, but once those are fixed I think this can be merged. I didn't test it locally, but since all tests pass this should be fine.

**Pull request merged**